Updated buttons to be more accessible.

dsjen opened the pull request:
## Description

[FEDX-231](https://openedx.atlassian.net/browse/FEDX-231)

Accessility updates to buttons:
* Darkened hover/focus colors
* Darked link read color
* Updated disabled button styling on inverse background

Preview: http://ux-test.edx.org/dsjen/a11y-buttons/components/buttons/
  
## Manual Test Plan

* Confirm color is darker when hovering over buttons.
* Confirm color of links is darker when "read."
* Confirm under link appears when hovering over links.
* Confirm disabled button is correct in inverse background state.

## Testing Checklist
- [ ] Manually test responsive behavior.
- [ ] Manually test right-to-left behavior.
- [ ] Manually test a11y support.

## Non-testing Checklist
- [ ] Consider any documentation your change might need, and which users will be affected by this change.

## Post-review
- [ ] Squash commits into discrete sets of changes with descriptive commit messages.

## Reviewers
- [ ] @chris-mike 
- [ ] @alisan617 
- [ ] @andy-armstrong 
- [x] @bjacobel 
- [ ] @thallada 

I'm looking for two developer 👍  in addition to @chris-mike as product reviewer.  If you've been tagged for review, please check your corresponding box once you've given the :+1:.

@chris-mike -- Please confirm that this is what you were expecting.

andy-armstrong commented:
@dsjen This looks great! A couple of minor things I noticed:
 - the elevated buttons (which are odd at the best of times) have an odd hover color for the bottom bar
 - this wasn't your change, but the inverse disabled button appears to have a cut-and-paste error (it is labeled "base" and doesn't have the correct class)

thallada commented on pldoc/_components/buttons.md:
```diff
@@ -93,7 +93,7 @@ info:          Buttons should be used for performing actions within the edX envi
         <button type="button" class="btn-inverse btn-large">Inverse Large</button>
         <button type="button" class="btn-inverse btn-small">Inverse Small</button>
-        <button type="button" class="btn-inverse" disabled>Inverse Disabled</button>
+        <button type="button" class="btn" disabled>Base Disabled</button>
```
I think the problem with replacing the class in this example is that it is still possible to have an inverse button that is disabled and get the old style. Shouldn't we tweak the inverse disabled button styling?

I also think using the base disabled style here sticks out like a sore thumb compared to the other inverse buttons next to it.

dsjen replied:
Yes, I agree!  The disabled-ness of it draws a lot of attention.  @chris-mike -- what do you think?  Is there another color combination?

thallada commented on pattern-library/sass/patterns/_buttons.scss:
```diff
@@ -120,8 +114,8 @@ $btn-small-font-size: font-size(small);
+        background: $btn-focus-color;
         color: palette(primary, x-back);
     }
 
```
I know this wasn't part of this PR, but do we have a ticket for improving the disabled button text color? It's way too low of a contrast, I'm sure it would trip the accessibility checker.

dsjen commented:
@andy-armstrong -- thanks for bringing attention the hover outline.  @chris-mike -- do you have a style in mind to address the odd looking outline in the "elevated" button now that we've darkened the hover color?

Also, the inverse disabled button is what was asked for in the ticket, but I agree that it looks very odd...even like an error.  @chris-mike -- recommendation on the disabled button with inverse background?

bjacobel commented:
:thumbsup:

dsjen commented:
I've made a few changes that @chris-mike suggested:
* added text beneath link heading
* adjusted inverse disabled button styles

Please see the screenshot as I'm unable to deploy a sandbox:
![blue-inverse-link-text](https://cloud.githubusercontent.com/assets/5265058/20355345/0d68da5e-abef-11e6-9336-32d88935a7ea.png)

Outstanding questions:
* Generally, contrast on text for disabled buttons seems low.  Should we address this now or on a separate ticket?
* The elevated button on hover seem to suggest that it's depressed rather.  I did a [quick search on edx-platform](https://github.com/edx/edx-platform/search?utf8=%E2%9C%93&q=btn-elevated) and see no usage of it.  Can we remove it?  Do we care if it looks funny at this point?

bjacobel commented:
+1 for removing the elevated button, I don't think it's consistent with any of our existing branding/style.

andy-armstrong commented:
+1 for removing the elevated buttons from me too. I don't see any usage of them anywhere.

@talbs can you give us the history on elevated buttons?

dsjen commented:
In response to our meeting this afternoon, I've updated this PR with the following:
* darkened disabled button text
* removed elevated buttons entirely

![screen shot 2016-11-16 at 3 44 23 pm](https://cloud.githubusercontent.com/assets/5265058/20365113/b5c81588-ac13-11e6-99b8-a60236c13018.png)

andy-armstrong commented on pattern-library/sass/patterns/_buttons.scss:
```diff
@@ -120,8 +120,8 @@ $btn-small-font-size: font-size(small);
     &.is-focused {
-        border-color: palette(primary, accent);
-        background: palette(primary, accent);
+        border-color: $btn-focus-color;
```
Thanks for making these into variables! We should strive never to use ``palette`` directly in our Sass, but instead use variables. This allows a theme to override the value for the variables and get a consistent experience.

andy-armstrong commented on pattern-library/sass/patterns/_buttons.scss:
```diff
@@ -120,8 +120,8 @@ $btn-small-font-size: font-size(small);
-        background: palette(primary, accent);
+        border-color: $btn-focus-color;
+        background: $btn-focus-color;
         color: palette(primary, x-back);
```
I'd understand if you see this as out of scope, but would you mind making this a variable too?

dsjen replied:
Sure thing!  Updating this to use `$btn-text-focus-color`.

andy-armstrong commented on pattern-library/sass/patterns/_buttons.scss:
```diff
@@ -138,7 +138,7 @@ $btn-small-font-size: font-size(small);
     &.is-disabled {
         border-color: palette(grayscale, back);
-        color: palette(grayscale, back);
+        color: palette(grayscale, accent);
```
Can you make this a variable too.

andy-armstrong commented on pattern-library/sass/patterns/_buttons.scss:
```diff
@@ -340,13 +306,8 @@ $btn-small-font-size: font-size(small);
-            border-color: palette(grayscale, back);
-            color: palette(grayscale, back);
-        }
+        border-color: palette(grayscale, accent);
```
Can you make these use variables.

andy-armstrong commented on pldoc/_components/buttons.md:
```diff
@@ -102,10 +90,10 @@ info:          Buttons should be used for performing actions within the edX envi
 </div>
 
 <h3 class="example-set-hd">Links without visual styling</h3>
+<p>To indicate 'read' or 'visited' links in content areas, use color Primary Dark.</p>
```
I find myself confused by this. I read this as advice to the user of the Pattern Library, but this isn't something in their control. Maybe you can make it sound less like an option, such as

    Note: unread links use the primary color, while read/visited links use primary dark.

dsjen replied:
Makes sense!  FYI, @chris-mike

dsjen commented:
Added the following variables:
```
$btn-focus-color:                           palette(primary, dark) !default;
$btn-focus-text-color:                      palette(primary, x-back) !default;
$btn-disabled-text-color:                   palette(grayscale, accent) !default;
$btn-disabled-background-color:             palette(grayscale, x-back) !default;
$btn-disabled-border-color:                 palette(grayscale, back) !default;
```
